src/librustc/hir/map/mod.rs

```diff
@@ -757,6 +877,20 @@ impl<'hir> Map<'hir> {
         }
     }
 
+    // FIXME(@ljedrz): replace the NodeId variant
+    pub fn get_parent_by_hir_id(&self, id: HirId) -> HirId {
+        match self.walk_parent_nodes_by_hir_id(id, |node| match *node {
+            Node::Item(_) |
+            Node::ForeignItem(_) |
+            Node::TraitItem(_) |
+            Node::ImplItem(_) => true,
+            _ => false,
+        }, |_| false) {
+            Ok(id) => id,
+            Err(id) => id,
+        }
+    }
+
```

Comment:
This should just convert the `HirId` to a `NodeId` and call the regular `get_parent` method. Then convert the result back into a `NodeId`. You could also rename this to `get_parent_item`, which is a better name for the operation.